lightningd/pay.c

```diff
@@ -57,12 +57,76 @@ void payment_succeeded(struct lightningd *ld, struct htlc_out *hout,
 	hout->cmd = NULL;
 }
 
+/* Return false if permanent failure at the destination, true if
+ * retrying is plausible. */
+static bool report_routing_failure(struct subd *gossip,
+				   const struct wallet_payment *payment,
+				   const struct onionreply *failure)
+{
+	const tal_t *tmpctx;
+	u8 *channel_update = NULL;
+	u64 unused64;
+	u32 unused32;
+	enum onion_type failcode = fromwire_peektype(failure->msg);
+	u8 *gossip_msg;
+	const struct pubkey *route_nodes;
+	const struct short_channel_id *route_channels;
+	int origin_index;
+
+	/* Check if at destination. */
+	if (failure->origin_index >= tal_count(payment->route_nodes)) {
+		/* Is it permanent? */
+		if ((failcode & PERM) != 0)
+			return false;
```

Comment:
We should still hand to gossipd!

Reply:
But what will we hand to gossipd?  In particular, what outgoing channel will we report via `gossip_routing_failure` message?

From my reading of BOLT 4, below is the list of possible errors a final payee node can return:

1. `invalid_realm` - permanent non-node error, but this "should" only happen if we send a realm that the recipient is ignorant of, and that indicates a bug on our side as presumably future realm types (that old nodes might not recognize) should be enabled in a global features bit.  Arguably this should be treated a a node error, as apparently we think that node supports some new realm byte, but the node itself denies cognizance of that realm byte.
2. `temporary_node_failure` - node error, so outgoing channel ignored, no problem.
3. `permanent_node_failure` - node error, so outgoing channel ignored, no problem.
4. `required_node_feature_missing` - similar to `invalid_realm` (permanent non-node error, this time the receiver thinks we support some feature but we ourselves are ignorant of that feature and did not pay attention to the even feature bit).
5. `unknown_payment_hash` - permanent non-node error, but may happen for example if payee is c-lightning and the invoice expired before the payment reached it.
6.  `incorrect_payment_amount` - permanent non-node error, but is arguably a bug on our side where some roundoff or some other issue made us go beyond the limits of the invoice being paid.
7.  `final_expiry_too_soon` - temporary non-node error, but we can actually reuse the exact same route, we just need to adjust the CLTVs.
8.  `final_incorrect_cltv_expiry` - temporary non-node error, but maybe this indicates a bug on our side?
9.  `final_incorrect_htlc_amount` - temporary non-node error, similar to `final_incorrect_cltv_expiry`.

So for now, I think, it is sensible to report node-level errors on the final node to gossip, but I am uncertain what channel to deactivate for the non-node errors; non-node errors at the final destination seem to suggest a bug or disagreement between us and the final node.

I will fix the other review items for now but I want to know your opinion for the final node case in more detail first.